src/session.rs

```diff
@@ -0,0 +1,198 @@
+
+        Ok(Session {
+            path,
+            read_only,
+            content,
+        })
+    }
+
+    pub fn headers(&self) -> Result<HeaderMap> {
+        let mut headers: HeaderMap = HeaderMap::try_from(&self.content.headers)?;
+        let cookies = self
+            .content
+            .cookies
+            .values()
+            .filter(|c| !c.has_expired())
+            .map(|c| format!("{}={}", c.name, c.value))
+            .collect::<Vec<_>>()
+            .join("; ");
+        if !cookies.is_empty() {
+            headers.insert(COOKIE, HeaderValue::from_str(&cookies)?);
+        }
+        Ok(headers)
+    }
+
+    pub fn save_headers(&mut self, request_headers: &HeaderMap) -> Result<()> {
+        for (key, value) in request_headers.iter() {
+            let key = key.as_str();
+            if key != "cookie" && !key.starts_with("content-") && !key.starts_with("if-") {
+                self.content
+                    .headers
+                    .insert(key.into(), value.to_str()?.into());
+            }
+        }
+        Ok(())
+    }
+
+    pub fn save_auth(&mut self, request_headers: &HeaderMap) -> Result<()> {
+        if let Some(value) = request_headers.get(AUTHORIZATION) {
+            self.content
+                .headers
+                .insert("authorization".into(), value.to_str()?.into());
+        }
+        Ok(())
+    }
+
+    pub fn save_cookies(&mut self, response_headers: &HeaderMap) -> Result<()> {
+        for cookie in response_headers.get_all(SET_COOKIE) {
+            let parsed_cookie = Cookie::from_str(cookie.to_str()?)?;
+            self.content
+                .cookies
+                .insert(parsed_cookie.name.clone(), parsed_cookie);
+        }
+        Ok(())
+    }
+
+    pub fn persist(&self) -> Result<()> {
+        if !self.path.exists() || !self.read_only {
+            if let Some(parent_path) = self.path.parent() {
+                fs::create_dir_all(parent_path)?;
+            }
+            let mut session_file = fs::File::create(&self.path)?;
+            let formatter = serde_json::ser::PrettyFormatter::with_indent(b"    ");
+            let mut ser = serde_json::Serializer::with_formatter(&mut session_file, formatter);
+            self.content.serialize(&mut ser)?;
+            session_file.write_all(b"\n")?;
+        }
+        Ok(())
+    }
+}
+
+fn insert_cookie(headers: &mut HeaderMap, cookie: HeaderValue) {
+    if let Some(existing_cookie) = headers.get(COOKIE) {
+        let cookie = HeaderValue::from_str(&format!(
+            "{}; {}",
+            existing_cookie.to_str().unwrap(),
+            cookie.to_str().unwrap()
+        ))
+        .unwrap();
+        headers.insert(COOKIE, cookie);
+    } else {
+        headers.insert(COOKIE, cookie);
+    }
+}
+
+pub fn merge_headers(mut headers1: HeaderMap, headers2: HeaderMap) -> HeaderMap {
+    let mut current_key = None;
+    for (key, value) in headers2 {
+        current_key = key.or(current_key);
```

Comment:
Huh, why this? Would `let key = key.unwrap();` work?

I'm surprised `key` is an `Option`, in the [docs](https://docs.rs/http/0.2.4/http/header/struct.HeaderMap.html#impl-IntoIterator) it seems that it would be a plain `HeaderName`. Do you know why it does that?

Reply:
I think it is because each key can be associated with multiple values.

From https://docs.rs/http/0.2.4/src/http/header/map.rs.html#1758-1764
```
    /// Creates a consuming iterator, that is, one that moves keys and values
    /// out of the map in arbitrary order. The map cannot be used after calling
    /// this.
    ///
    /// For each yielded item that has `None` provided for the `HeaderName`,
    /// then the associated header name is the same as that of the previously
    /// yielded item. The first yielded item will have `HeaderName` set.
```

Also see https://docs.rs/http/0.2.4/http/header/struct.HeaderMap.html#method.drain

Reply:
Ahh, thank you. That makes sense.